sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -110,11 +121,30 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
         "Operator will receive unsafe rows as input but cannot process unsafe rows")
     }
     RDDOperationScope.withScope(sparkContext, nodeName, false, true) {
+      prepare()
       doExecute()
     }
   }
 
   /**
+   * Do the preparation for SparkPlan.
+   */
+  final def prepare(): Unit = {
+    doPrepare
+  }
+
+  /**
+   * Overridden by concrete implementations of SparkPlan. It is guaranteed to run before any
+   * `execute` of SparkPlan. This is helpful when we want to launch some background works, e.g.,
+   * `BroadcastHashJoin` uses it to broadcast asynchronously.
+   *
+   * This is lazy to make sure running doPrepare is called only once for each SparkPlan.
+   */
+  protected lazy val doPrepare: Unit = {
```

Comment:
it's really strange for subclasses to override a lazy val that is really a method call. I think instead this should be a normal `protected def doPrepare(): Unit` and we should have another `private val prepareCalled` somewhere to make sure we only call this once.